### ISSUE TITLE
Allow importing binary-format expressions

### ISSUE THREAD

**Issue by singpolyma** [dhall-lang/dhall-lang#464]
The standard doesn't specifically allow or forbid this, AFAICT, but `dhall-haskell` doesn't seem to allow it so I expect it is de facto not supported at the moment.  It would be really great if the standard specifically stated that an imported expression may be in either UTF-8 source form or CBOR binary form, and that import resolution should parse/decode as necessary.

**Comment by Nadrieril** [comment]
Is there a principled way to detect in which format a given file is withou trying to parse it ?

**Comment by singpolyma** [comment]
From local disk there is not.  However, a CBOR decode attempt of dhall source text will fail very fast in practise.

**Comment by Nadrieril** [comment]
That means that when decoding a CBOR file with errors, a CBOR decode error will silently be transformed into an illegible parse error ?

**Comment by singpolyma** [comment]
There are a few ways to make that case better: if the content fail CBOR parse but is also not valid UTF-8, then it is more likely to be bad CBOR than bad source text for example.  Also, the list of all possible first-octets for dhall+cbor is known, I would have to check if that could be helpful.

The *most* principled way would be for the standard to require we use self-describing CBOR: https://tools.ietf.org/html/rfc7049#section-2.4.5 for binary encoding, which would then be very easy to detect in any context.

**Comment by Nadrieril** [comment]
We could require an annotation on the import instead: `./file.dhallb as Binary`, though `as` is not the best word for this.
Or more simply discriminate on the file extension, which should cover 99% or actual use cases.

**Comment by Nadrieril** [comment]
Self-describing CBOR would be even better, but would break hash backward-compatibility

**Comment by singpolyma** [comment]
> Self-describing CBOR would be even better, but would break hash backward-compatibility

In transition, could specify to strip the tag before hashing to keep hashes the same.‎

**Comment by Nadrieril** [comment]
This may not be feasible when using an external CBOR library for the encoding.

**Comment by Nadrieril** [comment]
I retract this: is the library supports the self-description tag, then it probably supports removing it too

**Comment by f-f** [comment]
I have a higher-level question about this: what would be the use-case for importing a binary expression? (vs only allowing UTF8 text)

**Comment by Gabriel439** [comment]
Note that for remote imports you could use the `Accept`/`Content-Type` headers for the request/response with a content type such as `application/x.dhall+cbor` for CBOR

I also have a similar question to @f-f: What would be the motivation for doing this for files?  The reason I ask is that I'm not sure who or what would author local CBOR-encoded Dhall files other than the cache.  For remote expressions it can make sense to CBOR-encode the transferred expression for compression purposes, but not for local files.

**Comment by singpolyma** [comment]
> Note that for remote imports you could use the `Accept`/`Content-Type` headers for the request/response with a content type such as `application/x.dhall+cbor` for CBOR

Yes, we should absolutely eventually register some MIME types for exactly this kind of use. That's why I called out local files in response to the original question -- they take more work.

> I also have a similar question to @f-f: What would be the motivation for doing this for files?

I would primarily want to do it for URL imports, however:

* Supporting different formats for URL vs file imports would be inconsistent
* Supporting "dumb storage service" URL imports with unhelpful MIME types would be nice
* Even locally, decoding CBOR will be faster than parsing text (especially with newer implementations that have less-optimised parsers)‎
* "Local" content may actually be network-accessed in many cases (IPFS, NFS, local files proxy-fetched from a non-http-addressable private content store, etc) and so may benefit from the smaller size of the CBOR encoding

**Comment by Gabriel439** [comment]
I think the main thing for me is to try not to store the content type of in-band within the file.  The two places I can think of to put that information are:

* The file extension (i.e. `.dhall` vs `.dhallb`)
* Extended file attributes

**Comment by singpolyma** [comment]
> I think the main thing for me is to try not to store the content type of in-band within the file.

I think I'd be fine with file extension or heuristics-style approaches -- but I'm curious why you have an issue with the idea of an in-band magic number? They seem pretty common (the `file` command often relies on them) and in this case CBOR has a standardised one already that doesn't affect CBOR load semantics at all (will be safely ignored by any compliant decoder already).

**Comment by Gabriel439** [comment]
@singpolyma: Mainly for reasons similar to the concerns that @Nadrieril raised: if something goes wrong you have to guess whether it was a CBOR decoding failure or a parse failure.  If we have two ways to represent Dhall source code then I'd like the interpreter to be able to distinguish which one it is working with unambiguously.

**Comment by f-f** [comment]
I'm a bit concerned about this from the security point of view. I don't have an attack vector, but I'm worried that this might open up more.
The reason is that apart from "design flaws" (i.e. logic bugs in the standard that allow exploits) the only other potential source of exploits is the parser (the main danger here is that an interpreter might "misinterpret" malicious expressions leading to escalation, etc). 
Currently we have two well-specified parsers (the UTF-8 and the CBOR), but only the first one is "exposed to the world" (and for this reason receives most of the attention IMO). The CBOR parser only works when reading from the cache, and since the binary cache is "local" to the system, it's quite more trusted than "stuff on the internet".
So for this reason I'd consider the CBOR parser "not exposed to the world" at the moment, and I think that allowing CBOR imports will double the potential attack surface.
(This is not a new worry and some people came to me concerned about this - but all the above addressed their concern so far)

Another security aspect about this is "inspectability": one of Dhall's selling points is that you can import potentially malicious expression, but also that it's easy to inspect your dependencies (and this is quite relevant lately, with some fuss about vetting dependencies in the npm and ruby worlds) because they are just links. However if we allow this then some dependency you might be using might have a binary import, and while that's still "inspectable" it's much harder than just reading the source.

As far as I understand the main reason to allow importing of "binary objects" would be "performance, especially on network imports"
I'm pretty sure the improvement would be noticeable (we need benchmarks before we can tell), but I don't think the bottleneck is there at the moment. In fact, I think a somewhat more clever import strategy for "remote local" imports would have a much bigger effect: for example if you try to import [this expression](https://raw.githubusercontent.com/dhall-lang/dhall-kubernetes/haskell-implementation/types.dhall) you'll notice that it takes a loong time (`1m25s` for me with a fresh cache and a good network), because there are several thousands local imports in the overall tree - they are local to that expression but remote from our point of view, so we have to resolve them over the network (and that's why I call them "remote local")
I don't have ideas for this nor I'm sure it is fixable, but this is just to say that if this is the usecase then we should spend effort where the bottleneck is (which requires finding the bottlenecks first)

**Comment by Profpatsch** [comment]
> So for this reason I'd consider the CBOR parser "not exposed to the world" at the moment, and I think that allowing CBOR imports will double the potential attack surface.

More than double, because CBOR is a large superset of dhall, and languages might have CBOR implementations of differing quality. CBOR is quite a large standard compared to e.g. bencode.

It might make sense to allow implementations to prefer downloading CBOR representations via `Allow` headers, but for God’s sake don’t try “autodiscovery” of the format (that’s a security hole the size of an aircraft carrier).

**Comment by singpolyma** [comment]
> @singpolyma: Mainly for reasons similar to the concerns that @Nadrieril raised: if something goes wrong you have to guess whether it was a CBOR decoding failure or a parse failure.

This would not be the case with https://tools.ietf.org/html/rfc7049#section-2.4.5 -- either the data starts with exactly the bytes `"\xD9\xD9\xF7"` and is CBOR, or it does not and is UTF-8 source code.  There's no guessing with this method.

**Comment by Gabriel439** [comment]
I think the main thing I would like to see is a motivating use case for ingesting CBOR-encoded Dhall files directly.  I want to make sure that I better understand the requirements of the problem we're trying to solve.

**Comment by Gabriel439** [comment]
To clarify more what I mean by motivating use case: currently we are building Dhall around displacing YAML.  One of the key features of YAML is that it is human-readable and human-writable, so we should design around that.  The reason I want to better understand the use case is to see if this is helping us in our goal of displacing YAML since I want to reserve our complexity budget for that.

**Comment by winitzki** [comment]
Not sure this discussion is still relevant. We have now the `Bytes` type and imports `as Bytes`.
